### ISSUE TITLE
SC2048: Don't warn on unquoted element number get `${#arr[*]}`

### ISSUE THREAD

**Issue by AloisMahdal** [koalaman/shellcheck#2218]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2048
- My shellcheck version (`shellcheck --version` or "online"): 0.7.2
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash


bar() {
    test ${#Foo[*]} -gt 0 && printf ' %q' "${Foo[@]}"
}

main() {
    local Foo
    Foo+=("$1")
}

main "$@"
```

#### Here's what shellcheck currently says:

    / # shellcheck sc.sh 

    In sc.sh line 5:
        test ${#Foo[*]} -gt 0 && printf ' %q' "${Foo[@]}"
             ^--------^ SC2048: Use "${array[@]}" (with quotes) to prevent whitespace problems.

    For more information:
      https://www.shellcheck.net/wiki/SC2048 -- Use "${array[@]}" (with quotes) t...

#### Here's what I wanted or expected to see:

No warning.

I believe this is an Exception to SC2048:  having `#` there means I'm getting *number of elements*, therefore, while `${foo[*]}` is dangerous, `${#foo[*]}` is perfectly safe -- it is guarranteed to return a single integer.

### Notes

So far, this specific syntax hasn't been raising warning.  I always reckoned the lack of warning intentional for the above reason, but it looks like it might have been "fixed" by fe25a2b from yesterday.

Note that in my codebase, I learned to actually prefer using`*` and unquoted when getting number of elements, because it's more distinctive than the usual array extrapolation:

    "${foo[@]}"     # (A) actual array extrapolation
    "${#foo[@]}"    # (B) length get (safe, no SC now)
    ${#foo[@]}      # (C) length get (still safe)
    ${#foo[*]}      # (D) length get (still safe, more distinctive)

The form (B) is more "safeguarded"  -- and would not raise warning now.

Howeveer, it's pretty similar to the ubiquitous array extrapolation (A), hence it IMHO make sense to always use the form (D) -- "no quotes, asterisk" -- in order to highlight the difference.

**Comment by koalaman** [comment]
You're right, this was a straight forward bug. The test for `$*` checked that the expansion started with `*` which precluded `!` and `#`. The check for `[*]` looked at the modifiers after the variable name, and therefore didn't consider any leading modifiers. Thanks!
